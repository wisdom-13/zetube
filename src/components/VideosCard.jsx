import React from 'react';
import { useNavigate } from 'react-router-dom';
import { formatAgo } from '../util/date';


export default function VideosCard({ video, type }) {

  const { title, thumbnails, channelTitle, publishedAt } = video.snippet;

  const navigate = useNavigate();
  const link = `/videos/watch/${video.id}`
  const isList = type === 'list';

  return (
    <li
      className={isList ? 'flex gap-1 p-2' : ''}
      onClick={() => navigate(link, { state: { video } })}>
      <img className={isList ? 'w-60 mr-2' : 'w-full'} alt='{title}' src={thumbnails.medium.url}></img>
      <div>
        <h3 className='font-semibold my-2 line-clamp-2'>{title}</h3>
        <h5 className='text-sm opacity-80'>{channelTitle} / </h5>
        <h5 className='text-sm opacity-80'>{formatAgo(publishedAt, 'ko')}</h5>
      </div>
    </li>
  );
}