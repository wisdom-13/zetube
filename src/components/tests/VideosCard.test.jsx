import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import VideosCard from '../VideosCard'
import { MemoryRouter, Route, Routes, useLocation } from 'react-router-dom'
import { formatAgo } from '../../util/date'
import { fakeVideo as video } from '../../tests/videos'
import { withRouter } from '../../tests/utils'

describe('VideoCard', () => {
  const { title, channelTitle, publishedAt, thumbnails } = video.snippet;

  it('renders video item', () => {
    render(
      withRouter(<Route path='/' element={<VideosCard video={video} />} />)
    )

    const image = screen.getByRole('img');
    expect(image.src).toBe(thumbnails.medium.url);
    expect(image.alt).toBe(title);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(channelTitle)).toBeInTheDocument();
    expect(screen.getByText(formatAgo(publishedAt, 'ko'))).toBeInTheDocument();
  })

  it('navigates to detailed video page with wideo state when clicked', () => {
    function LocationStateDisplay() {
      return <pre>{JSON.stringify(useLocation().state)}</pre>
    }
    render(
      withRouter(
        <>
          <Route path='/' element={<VideosCard video={video} />} />
          <Route path={`/videos/watch/${video.id}`} element={<LocationStateDisplay />} />
        </>
      )
    )

    const card = screen.getByRole('listitem');
    userEvent.click(card)

    expect(screen.getByText(JSON.stringify({ video }))).toBeInTheDocument();
  })
})