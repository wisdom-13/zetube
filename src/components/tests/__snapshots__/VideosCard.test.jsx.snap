// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard renders grid type correctly 1`] = `
<li
  className=""
  onClick={[Function]}
>
  <img
    alt="title"
    className="w-full"
    src="http://image/"
  />
  <div>
    <h3
      className="font-semibold my-2 line-clamp-2"
    >
      title
    </h3>
    <h5
      className="text-sm opacity-80"
    >
      channelTitle
    </h5>
    <h5
      className="text-sm opacity-80"
    >
      방금
    </h5>
  </div>
</li>
`;

exports[`VideoCard renders list type correctly 1`] = `
<li
  className="flex gap-1 p-2"
  onClick={[Function]}
>
  <img
    alt="title"
    className="w-60 mr-2"
    src="http://image/"
  />
  <div>
    <h3
      className="font-semibold my-2 line-clamp-2"
    >
      title
    </h3>
    <h5
      className="text-sm opacity-80"
    >
      channelTitle
    </h5>
    <h5
      className="text-sm opacity-80"
    >
      방금
    </h5>
  </div>
</li>
`;
